pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/nativefst/FST.java

```diff
@@ -0,0 +1,438 @@
+
+
+/**
+ * This is a top abstract class for handling finite state automata. These
+ * automata are arc-based, a design described in Jan Daciuk's <i>Incremental
+ * Construction of Finite-State Automata and Transducers, and Their Use in the
+ * Natural Language Processing</i> (PhD thesis, Technical University of Gdansk).
+ */
+public abstract class FST implements Iterable<ByteBuffer> {
+  /**
+   * @param in The input stream.
+   * @param length Length of input to be read
+   * @return Reads remaining bytes upto length from an input stream and returns
+   * them as a byte array. Null if no data was read
+   * @throws IOException Rethrown if an I/O exception occurs.
+   */
+  protected static byte[] readRemaining(InputStream in, int length)
+      throws IOException {
+    byte[] buf = new byte[length];
+    int readLen;
+
+    readLen = in.read(buf, 0, length);
+
+    if (readLen == -1) {
+      return null;
+    }
+
+    return buf;
+  }
+
+  /**
+   * Wrapper for the main read function
+   */
+  public static FST read(InputStream stream)
+      throws IOException {
+    return read(stream, false, new DirectMemoryManager(FST.class.getName()));
+  }
+
+  /**
+   * A factory for reading automata in any of the supported versions.
+   *
+   * @param stream
+   *          The input stream to read automaton data from. The stream is not
+   *          closed.
+   * @return Returns an instantiated automaton. Never null.
+   * @throws IOException
+   *           If the input stream does not represent an automaton or is
+   *           otherwise invalid.
+   */
+  public static FST read(InputStream stream, boolean hasOutputSymbols, PinotDataBufferMemoryManager memoryManager)
+      throws IOException {
+    FSTHeader header = FSTHeader.read(stream);
+
+    switch (header._version) {
+      case ImmutableFST.VERSION:
+        return new ImmutableFST(stream, hasOutputSymbols, memoryManager);
+      default:
+        throw new IOException(
+            String.format(Locale.ROOT, "Unsupported automaton version: 0x%02x", header._version & 0xFF));
+    }
+  }
+
+  /**
+   * A factory for reading a specific FST subclass, including proper casting.
+   *
+   * @param stream
+   *          The input stream to read automaton data from. The stream is not
+   *          closed.
+   * @param clazz A subclass of {@link FST} to cast the read automaton to.
+   * @param <T> A subclass of {@link FST} to cast the read automaton to.
+   * @return Returns an instantiated automaton. Never null.
+   * @throws IOException
+   *           If the input stream does not represent an automaton, is otherwise
+   *           invalid or the class of the automaton read from the input stream
+   *           is not assignable to <code>clazz</code>.
+   */
+  public static <T extends FST> T read(InputStream stream, Class<? extends T> clazz, boolean hasOutputSymbols)
+      throws IOException {
+    FST fst = read(stream, hasOutputSymbols, new DirectMemoryManager(FST.class.getName()));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the resource returned by the *DirectMemoryManager* constructor in a try-finally block. Currently, there are execution paths that do not contain closure statements, for example, when FST.read throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).